### PR TITLE
En 6169 delegation sc deploy

### PR DESCRIPTION
**Pull request by iulianpascalau**
added the possibility to launch SCs at genesis time

### CODE REVIEW

File: genesis/parsing/smartContractParser.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+	return nil
+}
+
+func (scp *smartContractParser) checkForDuplicates() error {
+	for idx1 := 0; idx1 < len(scp.initialSmartContracts); idx1++ {
+		ib1 := scp.initialSmartContracts[idx1]
+		for idx2 := idx1 + 1; idx2 < len(scp.initialSmartContracts); idx2++ {
+			ib2 := scp.initialSmartContracts[idx2]
+			if ib1.Owner == ib2.Owner {
```
Comment (sasurobert) [inline genesis/parsing/smartContractParser.go:118]: what if elrond wants to deploy a set of contracts ? it should be not a problem - as you can do it with different nonces.

File: genesis/process/shardGenesisBlockCreator.go [hunk 2]
```diff
@@ -303,3 +319,71 @@ func createProcessorsForShard(arg ArgsGenesisBlockCreator) (*genesisProcessors,
...
+	code, err := getSCCodeAsHex(sc.GetFilename())
+	if err != nil {
+		return err
+	}
+
+	vmType := sc.GetVmType()
+	deployTxData := strings.Join([]string{code, vmType, codeMetadataHex}, "@")
+
+	tx := &transactionData.Transaction{
+		Nonce:     0,
```
Comment (sasurobert) [inline genesis/process/shardGenesisBlockCreator.go:368]: nonce according to owner nonce if he has more to deploy.

File: cmd/node/config/genesisSmartContracts.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+[
+  {
+    "owner": "erd1pajyy0hn3kw5s9dn5pf8980fdy2d0evklrj03699uxtl4k0fyudqrzxf8j",
+    "filename": "./config/genesisContracts/answer.wasm",
+    "vm-type": "0500"
+  }
+]
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/node/config/genesisSmartContracts.json:7]: done

File: process/smartContract/process.go [hunk 4]
```diff
@@ -491,7 +492,18 @@ func (sc *scProcessor) DeploySmartContract(
 
 	sc.txFeeHandler.ProcessTransactionFee(consumedFee, txHash)
 
-	log.Debug("SmartContract deployed")
+	scGenerated := make([]string, 0, len(vmOutput.OutputAccounts))
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/smartContract/process.go:495]: done

File: genesis/parsing/smartContractParser.go (hunk 1 above)
Comment (iulianpascalau) [reply genesis/parsing/smartContractParser.go:118]: Agree, removed the functions, allowed to do multiple deployments

File: genesis/process/shardGenesisBlockCreator.go (hunk 2 above)
Comment (iulianpascalau) [reply genesis/process/shardGenesisBlockCreator.go:368]: Agree, this is a must as an address can deploy multiple contracts

File: genesis/data/initialSmartContract.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+func (isc *InitialSmartContract) OwnerBytes() []byte {
+	return isc.ownerBytes
+}
+
+// SetOwnerBytes will set the owner address as raw bytes
+func (isc *InitialSmartContract) SetOwnerBytes(owner []byte) {
+	isc.ownerBytes = owner
+}
+
+// GetOwner returns the owner
```
Comment (miiu96) [inline genesis/data/initialSmartContract.go:21]: ...the smart contract owner address

File: genesis/interface.go [hunk 6]
```diff
@@ -40,3 +40,20 @@ type DelegationDataHandler interface {
 	GetValue() *big.Int
 	IsInterfaceNil() bool
 }
+
+// InitialSmartContractHandler represents the interface that describes the smart contract
+type InitialSmartContractHandler interface {
```
Comment (miiu96) [inline genesis/interface.go:45]: ... the initial smart contract

File: genesis/parsing/accountsParser.go [hunk 7]
```diff
@@ -187,11 +187,11 @@ func (g *Genesis) checkInitialAccount(initialAccount *data.InitialAccount) error
...
 }
 
-func (g *Genesis) checkForDuplicates() error {
-	for idx1 := 0; idx1 < len(g.initialAccounts); idx1++ {
-		ib1 := g.initialAccounts[idx1]
-		for idx2 := idx1 + 1; idx2 < len(g.initialAccounts); idx2++ {
-			ib2 := g.initialAccounts[idx2]
+func (ap *accountsParser) checkForDuplicates() error {
+	for idx1 := 0; idx1 < len(ap.initialAccounts); idx1++ {
+		ib1 := ap.initialAccounts[idx1]
```
Comment (miiu96) [inline genesis/parsing/accountsParser.go:192]: maybe ia (initial account)  instead of ib

File: genesis/parsing/accountsParser.go [hunk 8]
```diff
@@ -216,9 +216,9 @@ func (g *Genesis) StakedUpon(address string) *big.Int {
 }
 
 // DelegatedUpon returns the value that was delegated upon the provided address
-func (g *Genesis) DelegatedUpon(address string) *big.Int {
+func (ap *accountsParser) DelegatedUpon(address string) *big.Int {
 	delegated := big.NewInt(0)
-	for _, ib := range g.initialAccounts {
+	for _, ib := range ap.initialAccounts {
```
Comment (miiu96) [inline genesis/parsing/accountsParser.go:221]: also here

File: genesis/parsing/accountsParser.go [hunk 9]
```diff
@@ -205,8 +205,8 @@ func (g *Genesis) checkForDuplicates() error {
 }
 
 // StakedUpon returns the value that was staked upon the provided address
-func (g *Genesis) StakedUpon(address string) *big.Int {
-	for _, ib := range g.initialAccounts {
+func (ap *accountsParser) StakedUpon(address string) *big.Int {
+	for _, ib := range ap.initialAccounts {
```
Comment (miiu96) [inline genesis/parsing/accountsParser.go:209]: ia instead of ib

File: genesis/parsing/accountsParser.go [hunk 10]
```diff
@@ -248,8 +248,8 @@ func (g *Genesis) InitialAccountsSplitOnAddressesShards(
 	}
 
 	var addresses = make(map[uint32][]genesis.InitialAccountHandler)
-	for _, in := range g.initialAccounts {
-		address, err := g.pubkeyConverter.CreateAddressFromBytes(in.AddressBytes())
+	for _, in := range ap.initialAccounts {
```
Comment (miiu96) [inline genesis/parsing/accountsParser.go:251]: ia instead of in

File: genesis/data/initialSmartContract.go (hunk 5 above)
Comment (iulianpascalau) [reply genesis/data/initialSmartContract.go:21]: done

File: genesis/interface.go (hunk 6 above)
Comment (iulianpascalau) [reply genesis/interface.go:45]: done

File: genesis/parsing/accountsParser.go (hunk 7 above)
Comment (iulianpascalau) [reply genesis/parsing/accountsParser.go:192]: renamed

File: genesis/parsing/accountsParser.go (hunk 8 above)
Comment (iulianpascalau) [reply genesis/parsing/accountsParser.go:221]: done

File: genesis/parsing/accountsParser.go (hunk 9 above)
Comment (iulianpascalau) [reply genesis/parsing/accountsParser.go:209]: done

File: genesis/parsing/accountsParser.go (hunk 10 above)
Comment (iulianpascalau) [reply genesis/parsing/accountsParser.go:251]: done

**Pull request merged**